libs/vm/src/module.cpp

```diff
@@ -178,6 +179,88 @@ fixed_point::fp64_t toFixed64(VM * /* vm */, AnyPrimitive const &from)
   return Cast<fixed_point::fp64_t>(from);
 }
 
+Ptr<Fixed128> toFixed128(VM *vm, AnyPrimitive const &from)
+{
+  fixed_point::fp128_t fixed;
+  switch (from.type_id)
+  {
+  case TypeIds::Bool:
+  {
+    fixed = static_cast<fixed_point::fp128_t>(from.primitive.ui8);
+    break;
+  }
+  case TypeIds::Int8:
+  {
+    fixed = static_cast<fixed_point::fp128_t>(from.primitive.i8);
+    break;
+  }
+  case TypeIds::UInt8:
+  {
+    fixed = static_cast<fixed_point::fp128_t>(from.primitive.ui8);
+    break;
+  }
+  case TypeIds::Int16:
+  {
+    fixed = static_cast<fixed_point::fp128_t>(from.primitive.i16);
+    break;
+  }
+  case TypeIds::UInt16:
+  {
+    fixed = static_cast<fixed_point::fp128_t>(from.primitive.ui16);
+    break;
+  }
+  case TypeIds::Int32:
+  {
+    fixed = static_cast<fixed_point::fp128_t>(from.primitive.i32);
+    break;
+  }
+  case TypeIds::UInt32:
+  {
+    fixed = static_cast<fixed_point::fp128_t>(from.primitive.ui32);
+    break;
+  }
+  case TypeIds::Int64:
+  {
+    fixed = static_cast<fixed_point::fp128_t>(from.primitive.i64);
+    break;
+  }
+  case TypeIds::UInt64:
+  {
+    fixed = static_cast<fixed_point::fp128_t>(from.primitive.ui64);
+    break;
+  }
+  case TypeIds::Float32:
+  {
+    fixed = static_cast<fixed_point::fp128_t>(from.primitive.f32);
+    break;
+  }
+  case TypeIds::Float64:
+  {
+    fixed = static_cast<fixed_point::fp128_t>(from.primitive.f64);
+    break;
+  }
+  case TypeIds::Fixed32:
+  {
+    fixed = static_cast<fixed_point::fp128_t>(fixed_point::fp32_t::FromBase(from.primitive.i32));
+    break;
+  }
+  case TypeIds::Fixed64:
+  {
+    fixed = static_cast<fixed_point::fp128_t>(fixed_point::fp64_t::FromBase(from.primitive.i64));
+    break;
+  }
+  default:
+  {
+    fixed = 0;
+    // Not a primitive
+    assert(false);
+    break;
+  }
+  }  // switch
+
+  return Ptr<Fixed128>(new Fixed128(vm, fixed));
```

Comment:
I think it would be nice to add a few etch tests checking what actually happens when you try to convert each of these types into fp128. I'm also interested in what happens in the default case in release mode (where the assertion will be ignored); is it just going to silently do nothing? That seems problematic.

Reply:
Certainly check that every primitive can convert to a Fixed128 via tests. But If you get to the default case, you have a bad programming error which should be fixed.

Reply:
added tests in libs/vm-modules/tests/unit/fixed_point.cpp